### PR TITLE
IDE uses new visualization API

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

PR integrates new visualization API that uses method pointers instead of lambdas as visualization expressions. 

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: app/gui/docs/product/visualizations.md [hunk 1]
```diff
@@ -37,34 +37,33 @@ Visualizations can be displayed in the following ways:
...
-  is not connected, the visualization becomes an interactive widget allowing the
-  user to specify data. For example, a map visualization will allow the user to
-  manually pick locations. After each change, the new locations will be sent to
-  the output port. Under the hood, widgets are represented as nodes and their
-  code lines are assigned with a dedicated "visualization" metadata.
-  Visualizations generate expressions always in the form of `name = data`, where
-  data is a hardcoded data produced from the visualization. For example, when
-  user clicks the map to define locations, the data could be a string literal
-  containing locations encoded in JSON.
+- **Fullscreen** Visualization attached to node can grow (animate) to ocupy full
```
Comment (JaroslavTulach) [inline app/gui/docs/product/visualizations.md]:
```suggestion
- **Fullscreen** Visualization attached to node can grow (animate) to occupy full
```

File: app/gui/docs/product/visualizations.md [hunk 2]
```diff
@@ -37,34 +37,33 @@ Visualizations can be displayed in the following ways:
...
+  where you can write text mixed with visualizations (including widgets for an
+  interactive experience).
+
+- **Widgets** In this mode visualizations behave like nodes but do not display
+  expressions. They have one input and one output port. If the input port is
+  connected, the visualization displays its value and passes its to the output
+  port. In case it is not connected, the visualization becomes an interactive
+  widget allowing the user to specify data. For example, a map visualization
+  will allow the user to manually pick locations. After each change, the new
+  locations will be sent to the output port. Under the hood, widgets are
```
Comment (JaroslavTulach) [inline app/gui/docs/product/visualizations.md:61]: That almost sounds like "dynamic widgets" to me!

File: app/gui/docs/product/visualizations.md [hunk 3]
```diff
@@ -201,24 +200,11 @@ In particular:
   as visualizations. The superclass defines a default constructor and a set of
   utilities:
 
-  - #### Method `setPreprocessorCode(code)`
-    Set an Enso code which will be evaluated on the server-side before sending
-    data to visualization. If not called, a default unspecified code is used
+  - #### Method `setPreprocessor(module,method)`
+    Set an Enso method which will be evaluated on the server-side before sending
+    data to visualization. If not called, a default unspecified method is used
```
Comment (JaroslavTulach) [inline app/gui/docs/product/visualizations.md]: _default unspecified_? That's a strange formulation in a **specification** of the behavior ;-)

File: app/gui/docs/product/visualizations.md [hunk 4]
```diff
@@ -297,9 +283,8 @@ visualization you may use the `setPreprocessor` method). This code defines an
 Enso function, which will be run by the compiler on data the visualization is
 attached to. Only the results of this code will be sent to the GUI. In the case
 of the JSON input format, the result of the call should be a valid JSON string.
-The code will be evaluated in the context of the `Main` module in the project
-where visualization is defined - you may use any symbol defined or imported in
-that module.
+The code will be evaluated in the context of the module where the preprocessor
+method is defined - you may use any symbol defined or imported in that module.
```
Comment (JaroslavTulach) [inline app/gui/docs/product/visualizations.md:288]: How can one pass arguments to the method? #3655 needs to be able to invoke the Enso engine part of the visualization with different arguments.

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
#3655 needs an ability to invoke the Enso part of the visualization with arguments. How can one do that with `setPreprocessor(module, method)`?

File: app/gui/controller/double-representation/src/module.rs [hunk 5]
```diff
@@ -325,6 +325,14 @@ impl TryFrom<String> for QualifiedName {
     }
 }
 
+impl TryFrom<&String> for QualifiedName {
+    type Error = failure::Error;
+
+    fn try_from(text: &String) -> Result<Self, Self::Error> {
+        Self::from_text(text)
+    }
+}
```
Comment (farmaazon) [inline app/gui/controller/double-representation/src/module.rs]: Why is this needed? Usually, you should use `&str` instead `String` everywhere.

File: app/gui/controller/engine-protocol/src/language_server/types.rs [hunk 6]
```diff
@@ -651,10 +651,8 @@ pub type ExpressionId = Uuid;
 pub struct VisualisationConfiguration {
     #[allow(missing_docs)]
     pub execution_context_id: ContextId,
-    /// A qualified name of the module containing the expression which creates visualisation.
-    pub visualisation_module: String,
-    /// An enso lambda that will transform the data into expected format, i.e. `a -> a.json`.
-    pub expression:           String,
+    /// An enso function that will transform the data into expected format.
+    pub expression:           MethodPointer,
```
Comment (farmaazon) [inline app/gui/controller/engine-protocol/src/language_server/types.rs]: If this is a function, why the name remained `expression`?

File: app/gui/src/presenter/graph/visualization/manager.rs [hunk 7]
```diff
@@ -532,8 +500,8 @@ impl Manager {
 mod tests {
     use super::*;
 
+    use crate::model::module;
```
Comment (farmaazon) [inline app/gui/src/presenter/graph/visualization/manager.rs:508]:
```suggestion
    use crate::model::module;
    
```

File: app/gui/view/graph-editor/src/component/visualization/foreign/java_script/visualization.js [hunk 8]
```diff
@@ -29,56 +29,25 @@ export class Visualization {
...
-    getPreprocessorCode() {
-        return this.__preprocessorCode__
-    }
-
-    /**
-     * Set new preprocessor code.
-     *
-     * When visualization is attached to a node, each time a new value is produced from node,
-     * the preprocessor shall be invoked with it. Result such call shall be serialized and
-     * transported to visualization by invoking onDataReceived(data) method.
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/component/visualization/foreign/java_script/visualization.js:52]: I miss such detailed documentation in the new version.

We have a special documentation page for contributors wanting to create new visualizations, and these docs AFAIK refers to this file for specific documentation of JS methods they can use inside their visualizations.

File: app/gui/docs/product/visualizations.md (hunk 3 above)
Comment (mwu-tow) [reply app/gui/docs/product/visualizations.md]: Why? Specification explicitly opts-out from defining this. 
There'll be some code present, but custom visualization should not make any assumptions about it. Thanks to that, we can adjust it without breaking public API.

File: app/gui/docs/product/visualizations.md (hunk 4 above)
Comment (mwu-tow) [reply app/gui/docs/product/visualizations.md:288]: I'd expect that arguments are part of the expression, as the arguments change we update the expression.

File: app/gui/src/model/execution_context.rs [hunk 9]
```diff
@@ -216,6 +216,78 @@ pub struct LocalCall {
...
+// === QualifiedMethodPointer ===
+// ==============================
+
+/// A method pointer containing the qualified module and type names.
+#[derive(Clone, Debug, PartialEq)]
+pub struct QualifiedMethodPointer {
+    /// A module name containing the method.
+    pub module:          module::QualifiedName,
+    /// A type on which the method is defined.
+    pub defined_on_type: module::QualifiedName,
```
Comment (mwu-tow) [inline app/gui/src/model/execution_context.rs]: If this is a type, why it uses `module::QualifiedName` rather than `tp::QualifiedName`?

File: app/gui/docs/product/visualizations.md (hunk 4 above)
Comment (mwu-tow) [reply app/gui/docs/product/visualizations.md:288]: Ah, it was removed. Then some mechanism for arguments is necessary,

File: app/gui/view/graph-editor/src/builtin/visualization/java_script/histogram.js [hunk 10]
```diff
@@ -44,7 +43,7 @@ class Histogram extends Visualization {
 
     constructor(data) {
         super(data)
-        this.setPreprocessor('process_to_json_text', 'Standard.Visualization.Histogram')
+        this.setPreprocessor('Standard.Visualization.Histogram', 'process_to_json_text')
```
Comment (JaroslavTulach) [inline app/gui/view/graph-editor/src/builtin/visualization/java_script/histogram.js:46]: Is there any reason to revert the order of arguments?

File: app/gui/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js [hunk 11]
```diff
@@ -49,7 +49,7 @@ class ScatterPlot extends Visualization {
 
     constructor(data) {
         super(data)
-        this.setPreprocessor('process_to_json_text', 'Standard.Visualization.Scatter_Plot')
+        this.setPreprocessor('Standard.Visualization.Scatter_Plot', 'process_to_json_text')
```
Comment (JaroslavTulach) [inline app/gui/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js]: Here #3655 will need to pass the additional arguments.

File: app/gui/controller/engine-protocol/src/language_server/types.rs (hunk 6 above)
Comment (4e6) [reply app/gui/controller/engine-protocol/src/language_server/types.rs]: I left the same name to keep the backward compatibility

File: app/gui/view/graph-editor/src/component/visualization/foreign/java_script/visualization.js [hunk 12]
```diff
@@ -55,6 +63,18 @@ export class Visualization {
...
 
+    /**
+     * Set the arguments of the preprocessor function.
+     *
+     * @param arguments the arguments passed to the preprocessor function.
+     */
+    setArguments(...args) {
+        if (args !== this.__preprocessorArguments__) {
+            this.__preprocessorArguments__ = args
+            this.__emitPreprocessorChange__()
```
Comment (JaroslavTulach) [inline app/gui/view/graph-editor/src/component/visualization/foreign/java_script/visualization.js:83]: Ideally changing only the `setArguments` shall be a _lightweight_ operation. Something that invokes already existing code on the engine side, without re-parsing the code. But I guess this is provided _by definition_ when using method pointers...

File: engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeVisualizationsTest.scala [hunk 13]
```diff
@@ -2198,7 +2204,176 @@ class RuntimeVisualizationsTest
                 "Enso_Test.Test.Visualisation",
                 "Enso_Test.Test.Visualisation",
                 "incAndEncode"
+              ),
+              Vector("2", "3")
```
Comment (JaroslavTulach) [inline engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeVisualizationsTest.scala:2208]: What other types are supported? `Text`? Booleans `True`, `False`? Complex atoms? Function invocations?

**Review by JaroslavTulach — APPROVED** [review]
Thanks. This should be good enough for the [scatterplot visualization](https://github.com/enso-org/enso/pull/3655).

File: engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeVisualizationsTest.scala (hunk 13 above)
Comment (4e6) [reply engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeVisualizationsTest.scala:2208]: An argument can be an arbitrary Enso expression that can be interpreted in the context of the visualization module

File: app/gui/view/graph-editor/src/builtin/visualization/java_script/histogram.js (hunk 10 above)
Comment (4e6) [reply app/gui/view/graph-editor/src/builtin/visualization/java_script/histogram.js:46]: I like it more. Imagine calling a method by a fully qualified name `Foo.Bar.Baz.quux a b c`. The `setPreprocessor` method just splits the call into three sections (arguments are optional in this case)

File: app/gui/view/graph-editor/src/component/visualization/foreign/java_script/visualization.js (hunk 8 above)
Comment (4e6) [reply app/gui/view/graph-editor/src/component/visualization/foreign/java_script/visualization.js:52]: I updated the doc with more details and examples. Should be better now

File: app/gui/src/presenter/graph/visualization/manager.rs [hunk 14]
```diff
@@ -94,6 +75,7 @@ pub enum Notification {
 
 /// Describes the state of the visualization on the Language Server.
 #[derive(Clone, Debug, PartialEq)]
+#[allow(clippy::large_enum_variant)]
```
Comment (4e6) [inline app/gui/src/presenter/graph/visualization/manager.rs:78]: Clippy complains about the size of enum variant. I though that it is safe to ignore because most of the variants have comparable sizes (the size of `Visualization` struct). And the variant having two `Visualization` fields is the one that Clippy complains about.

File: app/gui/src/model/execution_context.rs [hunk 15]
```diff
@@ -216,6 +217,78 @@ pub struct LocalCall {
...
+    pub name:            String,
+}
+
+impl QualifiedMethodPointer {
+    /// Create a method pointer representing a module method.
+    pub fn module_method(module: module::QualifiedName, name: String) -> QualifiedMethodPointer {
+        QualifiedMethodPointer { module: module.clone(), defined_on_type: module.into(), name }
+    }
+    /// Tries to create a new method pointer from string components.
+    pub fn from_unqualified(
```
Comment (mwu-tow) [inline app/gui/src/model/execution_context.rs]: Is this "from unqualified" or from "from qualified text"?

Comment (4e6) [reply app/gui/src/model/execution_context.rs]: Agree, that is the right name for the method

File: app/gui/src/model/execution_context.rs [hunk 16]
```diff
@@ -216,6 +217,78 @@ pub struct LocalCall {
 
 
 
+// ==============================
+// === QualifiedMethodPointer ===
+// ==============================
+
+/// A method pointer containing the qualified module and type names.
+#[derive(Clone, Debug, PartialEq)]
```
Comment (MichaelMauderer) [inline app/gui/src/model/execution_context.rs]: Is there a reason not to also have `Eq`? (This will soon be a Clippy lint).

File: app/gui/view/graph-editor/src/builtin/visualization/java_script/histogram.js (hunk 10 above)
Comment (MichaelMauderer) [reply app/gui/view/graph-editor/src/builtin/visualization/java_script/histogram.js:46]: Please double-check that all calls to this have been update. This is a change that can easily cause issues as the arguments have the same type and this will only fail at runtime.

File: app/gui/src/model/execution_context.rs (hunk 16 above)
Comment (4e6) [reply app/gui/src/model/execution_context.rs]: Didn't know that it's necessary. I'll add one

**Comment by 4e6** [comment]
> Please double-check that all calls to this have been update. This is a change that can easily cause issues as the arguments have the same type and this will only fail at runtime.

I'm pretty sure I updated all of them. Unless grep failed me :sweat_smile:

File: app/gui/controller/engine-protocol/src/language_server/tests.rs [hunk 17]
```diff
@@ -473,8 +489,12 @@ fn test_execution_context() {
...
             "visualisationConfig" : {
                 "executionContextId"  : "00000000-0000-0000-0000-000000000000",
-                "visualisationModule" : "[Foo.Bar.Baz]",
-                "expression"          : "1 + 1"
+                "expression"          : {
+                    "module"        : "[Foo.Bar.Baz]",
+                    "definedOnType" : "[Foo.Bar.Baz]",
+                    "name"          : "foo"
+                },
+                "positionalArgumentsExpressions" : []
```
Comment (mwu-tow) [inline app/gui/controller/engine-protocol/src/language_server/tests.rs]: It'd be nice to have tests where this is not always empty.

File: app/gui/src/model/execution_context.rs [hunk 18]
```diff
@@ -216,6 +217,78 @@ pub struct LocalCall {
...
+
+/// A method pointer containing the qualified module and type names.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct QualifiedMethodPointer {
+    /// A module name containing the method.
+    pub module:          module::QualifiedName,
+    /// A type on which the method is defined.
+    pub defined_on_type: tp::QualifiedName,
+    /// A method name.
+    pub name:            String,
```
Comment (mwu-tow) [inline app/gui/src/model/execution_context.rs]: There are stronger types like `Identifier` or `NotReferentName` available. Consider using appropriate one, if it is not too much hassle.

Unless there are more syntactically complex names that can appear here?

File: app/gui/docs/product/visualizations.md [hunk 19]
```diff
@@ -201,24 +200,11 @@ In particular:
   as visualizations. The superclass defines a default constructor and a set of
   utilities:
 
-  - #### Method `setPreprocessorCode(code)`
-    Set an Enso code which will be evaluated on the server-side before sending
-    data to visualization. If not called, a default unspecified code is used
+  - #### Method `setPreprocessor(module,method,arguments)`
```
Comment (mwu-tow) [inline app/gui/docs/product/visualizations.md]: It should be clarified that `arguments` are vararg rather than a singular argument.

**Review by mwu-tow — APPROVED** [review]
